classes/class-wc-rest-connect-migration-flag-controller.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+if ( class_exists( 'WC_REST_Connect_Migration_Flag_Controller' ) ) {
+	return;
+}
+
+class WC_REST_Connect_Migration_Flag_Controller extends WC_REST_Connect_Base_Controller {
+	protected $rest_base = 'connect/migration-flag';
+
+	private function is_valid_state($state) {
+		$valid_states = [
+			WC_Connect_API_Constants::MIGRATION_STATE_NOT_STARTED,
+			WC_Connect_API_Constants::MIGRATION_STATE_STARTED,
+			WC_Connect_API_Constants::MIGRATION_STATE_COMPLETED,
+			WC_Connect_API_Constants::MIGRATION_STATE_FAILED
+		];
+		if (in_array($state, $valid_states, true)) {
+			return true;
+		}
+		return false;
+	}
+
+	public function post($request) {
+		$params = $request->get_json_params();
+		$migration_state = intval($params['migration_state']);
+		if (!$this->is_valid_state($migration_state)) {
+			$error = new WP_Error(
+				'invalid_migration_state',
+				__( 'Invalid migration state. Can not update migration state.', 'woocommerce-services' ),
+				array( 'status' => 400 )
+			);
+			return $error;
+		}
+
+		$result = WC_Connect_Options::update_option( 'wcshipping_migration_state', $migration_state );
+
+		return new WP_REST_Response( array( 'result' => $result ), 200 );
```

Comment:
I thought about returning `new WP_Error()` if the `$result` is `false`. At the same time, I think we can let it fail gracefully with a 200 with `result=>false`. The UI can figure out what we want to do with the failed `update_option` call. Thus, I used the latter approach.

We can look at error handling around installation and activation holistically in woocommerce shipping issue 134.

Reply:
> Caveat: update_option returns false if $value === $old_value

What about returning 200 if it's `true`, and 304 (not modified) for `false`?

Reply:
Nice, I appreciate the specificity. I think `304` is for "Not Changed" as in "there's nothing new, you don't need to redownload this thing" rather than "you requested a change but I didn't make it" though. What about one of the 5xx codes?

Reply:
This one is a bit tricky. `update_option()` returns `false` if it didn't update or if there is nothing to update. `false` here implies it either failed or it didn't do anything. 
For example: If it hits this condition, then it would be a valid 304: https://github.com/WordPress/wordpress-develop/blob/2132c7bacb1626af19737ffea0bcce9e42eee997/src/wp-includes/option.php#L903

Without being too specific and too complicated, I think I agree `500` is a good choice, if the `connect/migration-flag` failed to update, we need to know something is wrong, `500` would be a good one.

Reply:
Fixed in 73f0d9f6

Tested the error: 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/7361fba0-5045-4068-8fca-55dc15008517)

Reply:
Ooooh, I see! I guess the solution would be to compare the values before updating and then if they're different, only then update. This would make the `false`s returned by `update_option` only indicate actual failure.

I suppose we shouldn't run into these situations when the UI governs these flows so it should be OK to keep as it is.